File: packages/engine/src/catalogFormat.ts
// The catalog SHAPE now lives in @runoff/core; the engine re-exports it so
// existing `import type { CatalogFamily } from "@runoff/engine"` sites are
// unchanged, and owns the serialization below.
export type { CatalogFamily, CatalogTable } from "@runoff/core";
import type { CatalogFamily, CatalogTable } from "@runoff/core";

function familyHead(f: CatalogFamily): string {
  const gran = f.kind === "periodic" ? `periodic, ${f.granularity}` : "constant";
  const filed = f.filedPeriods.length ? `; filed: ${f.filedPeriods.join(", ")}` : "";
  return `${f.key} — "${f.label}" (${gran}${filed})`;
}

function tableLine(t: CatalogTable): string {
  const cols = t.columns.map((c) => `${c.name} ${c.type}`).join(", ");
  const counts = Object.entries(t.rowCounts);
  const total = counts.reduce((a, [, n]) => a + n, 0);
  const perPeriod = counts.length > 1 || (counts.length === 1 && counts[0][0] !== "")
    ? ` (${counts.map(([p, n]) => `${p}: ${n}`).join(", ")})`
    : "";
  return `  ${t.name}(${cols}) — ${total.toLocaleString("en-US")} rows${perPeriod}`;
}

/** ~2 lines per table; document families collapse to one annotated line. */
export function serializeCatalog(families: CatalogFamily[]): string {
  return families
    .map((f) => (f.queryable ? [familyHead(f), ...f.tables.map(tableLine)].join("\n") : `${familyHead(f)} — document, not queryable`))
    .join("\n");
}

File: backend/runoff_api/engine/catalog_format.py
"""Catalog serialization — statement-for-statement port of
packages/engine/src/catalogFormat.ts.

The catalog SHAPE lives in core (runoff_api.core.types.catalog); this module owns
the ~2-lines-per-table text rendering the run prompt consumes. Catalogs flow as
plain dicts with camelCase keys at runtime:
  CatalogFamily {"id","key","label","kind","granularity","queryable","tables","filedPeriods"}
  CatalogTable  {"name","columns","rowCounts"}
"""

from __future__ import annotations


def _family_head(f: dict) -> str:
    gran = f"periodic, {f['granularity']}" if f["kind"] == "periodic" else "constant"
    filed = f"; filed: {', '.join(f['filedPeriods'])}" if f["filedPeriods"] else ""
    return f"{f['key']} — \"{f['label']}\" ({gran}{filed})"


def _table_line(t: dict) -> str:
    cols = ", ".join(f"{c['name']} {c['type']}" for c in t["columns"])
    counts = list(t["rowCounts"].items())
    total = sum(n for _, n in counts)
    per_period = (
        f" ({', '.join(f'{p}: {n}' for p, n in counts)})"
        if len(counts) > 1 or (len(counts) == 1 and counts[0][0] != "")
        else ""
    )
    return f"  {t['name']}({cols}) — {total:,} rows{per_period}"


def serialize_catalog(families: list[dict]) -> str:
    """~2 lines per table; document families collapse to one annotated line."""
    return "\n".join(
        "\n".join([_family_head(f)] + [_table_line(t) for t in f["tables"]])
        if f["queryable"]
        else f"{_family_head(f)} — document, not queryable"
        for f in families
    )
